File: python/tfidf.py
#!/usr/bin/env python

from math import log
from collections import defaultdict


class tfidf:
    def __init__(self):
        self.documents = {}
        self.tokens = defaultdict(lambda: defaultdict(int))

    def add_document(self, tokens):
        document_id = len(self.documents)
        self.documents[document_id] = tokens
        for token in tokens:
            self.tokens[token][document_id] += 1
        return document_id

    def search(self, token):
        results = [{
            'document_id': document_id
        } for document_id in self.tokens[token].keys()]

        for result in results:
            document_id = result['document_id']
            tf = float(self.tokens[token][document_id]) / len(self.documents[
                document_id])
            idf = log(float(len(self.documents)) / len(self.tokens[token]))
            result['score'] = tf * idf
            result['body'] = self.documents[document_id]

        results.sort(reverse=True, key=lambda result: result['score'])
        return results

File: javascript/tfidf.js
var documentIndex = {};
var tokenIndex = {};

function addDocument(tokens) {
  var documentId = Object.keys(documentIndex).length;
  documentIndex[documentId] = tokens;
  tokens.forEach(function(token) {
    if (token in tokenIndex) {
      if (documentId in tokenIndex[token]) {
        tokenIndex[token][documentId] += 1;
      } else {
        tokenIndex[token][documentId] = 1;
      }
    } else {
      tokenIndex[token] = {};
      tokenIndex[token][documentId] = 1;
    }
  });
  return documentId;
}

function search(token) {
  var results = Object.keys(tokenIndex[token]).map(function(documentId) {
    return { 'documentId': documentId };
  });
  results.forEach(function(result) {
    tf = tokenIndex[token][result.documentId] / documentIndex[result.documentId].length;
    idf = Math.log(Object.keys(documentIndex).length / Object.keys(tokenIndex[token]).length);
    result['score'] = tf * idf;
    result['body'] = documentIndex[result.documentId];
  });

  results.sort(function(a, b) {
    return b.score - a.score;
  });
  return results;
}
